lib/src/git.rs

```diff
@@ -1395,53 +1509,157 @@ fn remove_remote_refs(mut_repo: &mut MutableRepo, remote_name: &str) {
+
+    if git_repo.try_find_remote(new_remote_name).is_some() {
+        return Err(GitRemoteManagementError::RemoteAlreadyExists(
+            new_remote_name.to_owned(),
+        ));
+    }
+
+    remote
+        .replace_refspecs(
+            [format!("+refs/heads/*:refs/remotes/{new_remote_name}/*").as_bytes()],
+            gix::remote::Direction::Fetch,
+        )
+        .expect("default refspec to be valid");
+    remote
+        .replace_refspecs::<&BStr>([], gix::remote::Direction::Push)
+        .expect("empty refspecs to be valid");
+
+    let mut config = git_repo.config_snapshot().clone();
+    remote
+        .save_as_to(new_remote_name, &mut config)
+        .map_err(GitRemoteManagementError::from_git)?;
+    remove_remote_git_config_sections(&mut config, old_remote_name);
+    save_git_config(&config).map_err(GitRemoteManagementError::GitConfigSaveError)?;
+
+    rename_remote_git_refs(&mut git_repo, old_remote_name, new_remote_name)
+        .map_err(GitRemoteManagementError::from_git)?;
+
     if old_remote_name != REMOTE_NAME_FOR_LOCAL_GIT_REPO {
         rename_remote_refs(mut_repo, old_remote_name, new_remote_name);
     }
+
     Ok(())
 }
 
+fn rename_remote_git_refs(
+    git_repo: &mut gix::Repository,
+    old_remote_name: &str,
+    new_remote_name: &str,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    let old_prefix = format!("refs/remotes/{old_remote_name}/");
+    let new_prefix = format!("refs/remotes/{new_remote_name}/");
+    let ref_log_message = BString::from(format!(
+        "renamed remote {old_remote_name} to {new_remote_name}"
+    ));
+
+    git_repo.edit_references(
+        git_repo
+            .references()?
+            .prefixed(old_prefix.clone())?
+            .map(|result| {
+                result.map(|old_ref| {
+                    let mut new_name = BString::from(new_prefix.as_bytes());
+                    new_name.extend_from_slice(
+                        old_ref
+                            .name()
+                            .as_bstr()
+                            .strip_prefix(old_prefix.as_bytes())
+                            .expect("old ref name to have old prefix"),
+                    );
+                    [
+                        add_ref(
+                            new_name.try_into().expect("new ref name to be valid"),
+                            old_ref.target().into_owned(),
+                            ref_log_message.clone(),
+                        ),
+                        remove_ref(old_ref),
+                    ]
+                })
+            })
+            .flatten_ok()
+            .collect::<Result<Vec<_>, _>>()?,
+    )?;
+    Ok(())
+}
+
+/// Set the `url` to be used when fetching data from a remote.
+///
+/// Shim for the missing `gix::Remote::fetch_url` API.
+///
+/// **TODO:** Upstream an implementation of this to `gix`.
+fn remote_fetch_url<Url, E>(
```

Comment:
nit: rename to `remote_with_fetch_url`? `remote_fetch_url` sounds like returning a URL.

Reply:
Done.